server/client.go

```diff
@@ -853,10 +1118,21 @@ func (c *client) unsubscribe(sub *subscription) {
 		return
 	}
 	c.traceOp("<-> %s", "DELSUB", sub.sid)
+
 	delete(c.subs, string(sub.sid))
 	if c.srv != nil {
 		c.srv.sl.Remove(sub)
 	}
+
+	// If we are a queue subscriber on a client connection and we have routes,
+	// we will remember the remote sid and the queue group in case a route
+	// tries to deliver us a message. Remote queue subscribers are directed
+	// so we need to know what to do to avoid unnecessary message drops
+	// from [auto-]unsubscribe.
+	if c.typ == CLIENT && c.srv != nil &&
+		len(sub.queue) > 0 && c.srv.NumRoutes() > 0 {
```

Comment:
That's a big no-no. NumRoutes() grabs the server lock. We can't do that. We have code doing s.mu.Lock() -> c.mu.Lock() so we should not do the opposite otherwise we have the risk of getting a lock inversion.
If you really want to optimize, it would be with an atomic (or immutable depending of config reload capabilities) on if routing is enabled or not. But if it is, I would not check the number of routes, simply store the q group.

Reply:
Good catch, I will snapshot the variables and release client lock.